### PR TITLE
C++: DefaultTaintTracking perf fix for globals

### PR DESCRIPTION
**Pull request by jbj**
This fixes performance of #2491 on awslabs/s2n and hopefully other projects. See my messages starting at https://github.slack.com/archives/CP0LHP150/p1580812298253100.

I'd like to run all the security tests to confirm that there are no changes from this PR. We can use this feature in the IR virtual dispatch library too, but I don't think it necessarily has to be in this PR.

I recommend reading the individual commits.

### CODE REVIEW

File: cpp/ql/test/library-tests/dataflow/dataflow-tests/IRDataflowTestCommon.qll [hunk 1]
```diff
@@ -36,9 +36,27 @@ class TestAllocationConfig extends DataFlow::Configuration {
...
+  }
+
   override predicate isBarrier(DataFlow::Node barrier) {
     barrier.asExpr().(VariableAccess).getTarget().hasName("barrier")
   }
 
   override predicate isBarrierGuard(DataFlow::BarrierGuard bg) { bg instanceof TestBarrierGuard }
 }
+
+private predicate readsVariable(LoadInstruction load, Variable var) {
```
Comment (geoffw0) [inline cpp/ql/test/library-tests/dataflow/dataflow-tests/IRDataflowTestCommon.qll]: It's a shame we have to duplicate code (from DefaultTaintTracking.qll) here.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll [hunk 2]
```diff
@@ -219,10 +240,41 @@ class DefinitionByReferenceNode extends Node {
   }
 }
 
+/**
+ * A `Node` corresponding to a variable in the program, as opposed to the
+ * value of that variable at some particular point. This can be used for
+ * modelling flow in and out of global variables.
```
Comment (geoffw0) [inline cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll]: Is there value or cost to the fact that we also model non-global variables here?

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll]: Value: yes. Callers might use it to crudely overapproximate flow through fields, or they might at least use it for flow through `static` fields. It's only for the sake of bug-for-bug backwards compatibility that I haven't added flow through static fields in `DefaultTaintTracking`.

Cost: not much. It adds a small number of tuples to the IPA type -- the number of variables in the program, which should be much less than the number of instructions.

File: cpp/ql/test/library-tests/dataflow/dataflow-tests/IRDataflowTestCommon.qll (hunk 1 above)
Comment (jbj) [reply cpp/ql/test/library-tests/dataflow/dataflow-tests/IRDataflowTestCommon.qll]: The duplication might look less bad if we just in-line it.

Comment (geoffw0) [reply cpp/ql/test/library-tests/dataflow/dataflow-tests/IRDataflowTestCommon.qll]: Yes, and it isn't a lot of code (I thought we had a mechanism for testing things that are `private`?)

Comment (jbj) [reply cpp/ql/test/library-tests/dataflow/dataflow-tests/IRDataflowTestCommon.qll]: Quick Evaluation can now peek through private items, but qltest can't. The standard workaround is to make the private items public in a public `module Internal`, and then users know they're asking for trouble if they access anything in that module.

I can also delete the change to the tests in this directory. I wrote them for my own debugging purposes, but they are probably redundant now that the code is working, and there are better tests in the `DefaultTaintTracking` dir.

**Comment by jbj** [comment]
I think this PR is ready for review now. I've verified that it causes no result changes when enabling `DefaultTaintTracking`, and I've tested performance on awslabs/s2n.

File: cpp/ql/test/library-tests/dataflow/dataflow-tests/IRDataflowTestCommon.qll (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/test/library-tests/dataflow/dataflow-tests/IRDataflowTestCommon.qll]: I'm happy with inlining it.

**Comment by geoffw0** [comment]
I've just tried out this branch, it seemed to help significantly on my cppcheck snapshot (while everything else I tried wobbled around).  Results were all the same.  :+1:

**Pull request merged**